### DISCUSSION TITLE
Is there examples about using multiple Inputs observations?

### DISCUSSION THREAD

**Discussion by akux2021** [Toni-SM/skrl#205] [Q&A]
Hi @Toni-SM ,

Are there examples of using multiple Inputs observations?

For example, one input is an image and another input is a vector.

Something similar in stablebaseline3 here https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html#multiple-inputs-and-dictionary-observations

**Comment by Toni-SM** [comment]
Hi @akux2021 

Yes, skrl allows using complex spaces like `gym.spaces.Dict`.

The observations are automatically converted to flatten tensors to store them in memory. 
The flattened tensor is forward to the models `.compute()` method.
Models have a method to convert the flattened tensor to the original spaces: [`Model.tensor_to_space`](https://skrl.readthedocs.io/en/latest/modules/skrl.models.base_class.html#skrl.models.torch.base.Model.tensor_to_space). 

You can see the method documentation example code for more details. 
Also, you can see its usage in the DeepMind's [reach_site_vision](https://skrl.readthedocs.io/en/latest/intro/examples.html#learning-in-a-deepmind-environment) example. Specifically, this feature is highlighted and commented in this example

**Comment by akux2021** [comment]
Thanks, it`s very helpful!

**Comment by elle-miller** [comment]
Hi @Toni-SM, I am having difficulty with using dictionary observation spaces. I am using SKRL with Isaac Lab environments.

Where/how are the dictionary observations flattened? I have looked through the code but can't find evidence of this. At the moment the observation is arriving as a dictionary to `compute()`, and not a flattened tensor.

This would be fine, except I am getting errors since the memory class expects flattened tensors for storing. Would it be possible for the memory class to store dictionaries?

Specifically, the error comes from `add_samples()` in `Memory`
```
# dimensions and shapes of the tensors (assume all tensors have the dimensions of the first tensor)
tmp = tensors.get("states", tensors[next(iter(tensors))])  # ask for states first
dim, shape = tmp.ndim, tmp.shape
...
AttributeError: 'dict' object has no attribute 'ndim'
```

**Comment by Toni-SM** [comment]
Hi @elle-miller 

Due to Isaac Lab that only supports `Box` spaces at the moment, the automatic conversion from space to tensor is not implemented in the wrapper (in contrast to the gym/gymnasium wrapper). This is why the error is generated when trying to store the information in memory.

I am working (on the [toni/space_tensor_conversion](https://github.com/Toni-SM/skrl/tree/toni/space_tensor_conversion) branch) on support it... If you can wait, I can have a working version by the end of next week :sweat_smile:.

Btw, are you using the skrl integration in Isaac Lab (`skrl_ppo_cfg.yaml`) or the skrl's [Python scripts examples](https://skrl.readthedocs.io/en/latest/intro/examples.html#isaac-lab-environments)?

**Comment by elle-miller** [comment]
@Toni-SM Thanks very much, that is great to hear! Looking forward to the new version :smile: 

I am using the skrl integration in Isaac Lab `skrl_ppo_cfg.yaml`

**Comment by Toni-SM** [comment]
Hi @elle-miller Support for different observation spaces should be work now with the [toni/space_tensor_conversion](https://github.com/Toni-SM/skrl/tree/toni/space_tensor_conversion) branch

You can install and try it via

```
./isaaclab.sh -p -m pip install git+https://github.com/Toni-SM/skrl.git@toni/space_tensor_conversion
```

**Comment by btabia** [comment]
Hello, 

I am trying to use Dict for the same usage as above (Double CNN + MLP) for processing image through a feature extractor and scalar data through a MLP but I keep having this error. 

I am using only a Gymnasium environment. Therefore, how dict shall be use for passing observations to the model?

Many Thanks

`Error executing job with overrides: []
Traceback (most recent call last):
  File "", line 154, in main
    trainer = SequentialTrainer(cfg=cfg_trainer, env=env, agents=[model])
  File "", line 55, in __init__
    self.agents.init(trainer_cfg=self.cfg)
  File "skrl/agents/torch/ppo/ppo.py", line 187, in init
    self.memory.create_tensor(name="states", size=self.observation_space, dtype=torch.float32)
  File "site-packages/skrl/memories/torch/base.py", line 197, in create_tensor
    size = self._get_space_size(size, keep_dimensions)
  File "site-packages/skrl/memories/torch/base.py", line 122, in _get_space_size
    raise ValueError("keep_dimensions=True cannot be used with Dict spaces")
ValueError: keep_dimensions=True cannot be used with Dict spaces`
